### PR TITLE
docs: Update docs for kcrps.

### PR DESCRIPTION
**Pull request by jakob-schloer**
## Description

With kcrps we brought in a significant amount of changes to anemoi-training and models. This is my first attempt to update the documentations to describe these changes.

## Type of Change
-   [x] Documentation update

Happy to get suggestions and contributions to this.

----
📚 Documentation preview 📚: https://anemoi-training--258.org.readthedocs.build/en/258/

----
📚 Documentation preview 📚: https://anemoi-graphs--258.org.readthedocs.build/en/258/

----
📚 Documentation preview 📚: https://anemoi-models--258.org.readthedocs.build/en/258/

### CODE REVIEW

File: training/docs/user-guide/distributed.rst [hunk 1]
```diff
@@ -45,10 +45,37 @@ number of GPUs you wish to shard the model across. It is recommended to
 only shard if the model does not fit in GPU memory, as data distribution
 is a much more efficient way to parallelise the training.
 
+Anemoi Training provides different sharding strategies depending if the
+model task is deterministic or ensemble based.
+
+For deterministic models, the ``DDPGroupStrategy`` is used: .. code::
```
Comment (JPXKQX) [inline training/docs/user-guide/distributed.rst]:
```suggestion
For deterministic models, the ``DDPGroupStrategy`` is used:

  .. code::
```

File: training/docs/user-guide/models.rst [hunk 2]
```diff
@@ -80,4 +98,43 @@ coarser than the resolution of the base data.
...
 The encoder and decoder can be chosen to be a GNN or a GraphTransformer.
 This choice is independent of the processor, but currently the encoder
-and decoder must be the same model type otherwise the code will break,
+and decoder must be the same model type otherwise the code will break.
+
+******************
+ Field Truncation
+******************
+
+Field truncation is a post-processing step applied during autoregressive
```
Comment (JPXKQX) [inline training/docs/user-guide/models.rst]: is it post or pre?

File: training/docs/user-guide/training.rst [hunk 3]
```diff
@@ -289,6 +316,27 @@ locations change in time.
...
+
+   training_loss:
+      _target_: anemoi.training.losses.mse.WeightedMSELoss
+      # class kwargs
+
+The choice of loss function depends on the model task and the desired
+properties of the forecast.
+
+For ensemble training, the following loss functions are available: -
+**Kernel CRPS**: Continuous Ranked Probability Score using kernel
```
Comment (JPXKQX) [inline training/docs/user-guide/training.rst]: It doesn't display as a list

File: training/src/anemoi/training/losses/weightedloss.py [hunk 4]
```diff
@@ -37,8 +37,7 @@ def __init__(
 
         Exposes:
         - self.avg_function: torch.nanmean or torch.mean
-        - self.sum_function: torch.nansum or torch.sum
-        depending on the value of `ignore_nans`
+        - self.sum_function: torch.nansum or torch.sum depending on the value of `ignore_nans`
```
Comment (JPXKQX) [inline training/src/anemoi/training/losses/weightedloss.py]: I would remove the changes in this file as they don't seem to be big, and we can avoid conflicts when merging the loss refactor PR

**Review by JPXKQX — COMMENTED** [review]
Hi Jakob, thanks for updating the doc! I think from a user's point of view, it would be really useful to have a tutorial/user guide named "From deterministic to probabilistic" explaining what needs to be changed when moving from one scenario to the other.

File: models/docs/introduction/overview.rst [hunk 5]
```diff
@@ -119,3 +125,63 @@ The `distributed` module provides utilities for distributed training of
...
+***************
+ Layer Kernels
+***************
+
+Layer kernels provide a flexible mechanism to customize the
+implementation of linear layers and layer normalization in different
+parts of the model (encoder, processor, decoder) through the
+`config.yaml`.
+
+Example configuration: .. code-block:: yaml
```
Comment (JPXKQX) [inline models/docs/introduction/overview.rst]:
```suggestion
Example configuration: 

.. code-block:: yaml
```

File: models/docs/introduction/overview.rst [hunk 6]
```diff
@@ -119,3 +125,63 @@ The `distributed` module provides utilities for distributed training of
...
+
+   layer_kernels:
+      processor:
+         LayerNorm:
+            _target_: torch.nn.LayerNorm _partial_: True # Any arguments
```
Comment (JPXKQX) [inline models/docs/introduction/overview.rst]: should all `_partial_` should go to the next line?

File: models/docs/modules/models.rst [hunk 7]
```diff
@@ -2,40 +2,82 @@
...
 *********************************
- Encoder Processor Decoder Model
+ Encoder-Processor-Decoder Model
 *********************************
 
-The model defines a graph neural network architecture with configurable
-encoder, processor, and decoder.
+The model defines a network architecture with configurable encoder,
+processor, and decoder components.
+
```
Comment (ssmmnn11) [inline models/docs/modules/models.rst:16]: Add reference here to aifs pre-print? 

The model defines a network architecture with configurable encoder,
processor, and decoder components (e.g. Lang et al 2024a).

https://arxiv.org/html/2412.15832v1

File: models/docs/modules/models.rst [hunk 8]
```diff
@@ -2,40 +2,82 @@
...
+   :members:
+   :no-undoc-members:
+   :show-inheritance:
+
+******************************************
+ Ensemble Encoder-Processor-Decoder Model
+******************************************
+
+The ensemble model architecture implementing the AIFS-CRPS approach
+(reference: Lang et al. 2024).
```
Comment (ssmmnn11) [inline models/docs/modules/models.rst]: could  be Lang et al. 2024b, if we mention the det refernce above.

File: models/docs/introduction/overview.rst [hunk 9]
```diff
@@ -55,49 +55,55 @@ Currently the package includes the following preprocessing steps:
...
 to process the input data.
 
 The layers are designed as extensible classes to allow for easy
 experimentation and switching out of components.
 
+Mappers
+=======
+
 The layers implement `Mappers`, which can encode and decode data from
 the graph representation. The `Mappers` are used to process the input
```
Comment (ssmmnn11) [inline models/docs/introduction/overview.rst]: not in scope here, but would be better to describe everything with respect to 

input grid
internal grid / hidden grid
output grid

the graphs then describe via the edges how to connect these.

File: models/docs/introduction/overview.rst [hunk 10]
```diff
@@ -55,49 +55,55 @@ Currently the package includes the following preprocessing steps:
...
 data and output the prediction. The `Mappers` use the `Blocks` to
 process the data, which are the building blocks of the graph neural
 network.
 
+Processors
+==========
+
 Additionally, the layers implement `Processors` which are used to
 process the data in the graph representation. The `Processors` use a
```
Comment (ssmmnn11) [inline models/docs/introduction/overview.rst]: Again, out of scope but: Processors process the data on the internal grid. The transformer has no edge information for example.

File: training/docs/user-guide/models.rst (hunk 2 above)
Comment (ssmmnn11) [reply training/docs/user-guide/models.rst]: pre :-)

File: training/docs/user-guide/models.rst [hunk 11]
```diff
@@ -80,4 +98,43 @@ coarser than the resolution of the base data.
...
+rollout. It smooths the model output before feeding it back as input for
```
Comment (ssmmnn11) [inline training/docs/user-guide/models.rst]: It smooths part of the input data.

File: training/docs/user-guide/training.rst [hunk 12]
```diff
@@ -41,9 +41,23 @@ To configure the training:
...
+ Parallelization
+*****************
+
+Anemoi Training supports different parallelization strategies based on
+the training task (see :ref:`Strategy <strategy target>`):
+
+-  **DDPGroupStrategy**: Used for deterministic training tasks
+-  **DDPEnsGroupStrategy**: Used for ensemble training tasks
+
+These strategies are automatically selected based on the model task
```
Comment (ssmmnn11) [inline training/docs/user-guide/training.rst]: I think we need to do this manually.

File: training/docs/user-guide/training.rst [hunk 13]
```diff
@@ -137,14 +151,27 @@ will be classed as a prognostic variable.
...
+**************
+ Data Modules
+**************
+
+Anemoi Training provides different data modules to handle various model
+tasks:
+
+-  **AnemoiDatasetDataModule**: Standard data module for deterministic
+   training
+-  **AnemoiEnsDatasetsDataModule**: Specialized data module for ensemble
```
Comment (ssmmnn11) [inline training/docs/user-guide/training.rst:164]: **AnemoiEnsDatasetsDataModule**: Specialized data module for ensemble
   training. It also allows for training with perturbed initial conditions.

**Review by ssmmnn11 — CHANGES_REQUESTED** [review]
great work! some suggestions :-)

File: models/docs/introduction/overview.rst [hunk 14]
```diff
@@ -55,49 +55,55 @@ Currently the package includes the following preprocessing steps:
...
-The model is designed to take in a graph representation of the input
-data, process it using a series of modules, and output a prediction.
+The package currently includes the following model architectures:
 
-Currently the package includes the following models:
+-  **AnemoiModelEncProcDec**: The base encoder-processor-decoder
+   architecture (AIFS-single)
+-  **AnemoiModelEncProcDecHierarchical**: A hierarchical variant of the
+   base architecture
+-  **AnemoiEnsModelEncProcDec**: The ensemble version (AIFS-CRPS) that
```
Comment (HCookie) [inline models/docs/introduction/overview.rst]: Might be a good idea to remove the reference to the AIFS here, and just leave it is a score optimised ensemble model?

File: models/docs/introduction/overview.rst [hunk 15]
```diff
@@ -119,3 +125,63 @@ The `distributed` module provides utilities for distributed training of
...
+
+         Linear:
+            _target_: torch.nn.Linear _partial_: True
+
+      decoder:
+         LayerNorm:
+            _target_: torch.nn.LayerNorm _partial_: True
+
+         Linear:
+            _target_: torch.nn.Linear _partial_: True
```
Comment (HCookie) [inline models/docs/introduction/overview.rst]:
```suggestion
   layer_kernels:
      processor:
         LayerNorm:
            _target_: torch.nn.LayerNorm 
            _partial_: True 
            # Any arguments to your chosen function go here

         Linear:
            _target_: torch.nn.Linear 
            _partial_: True 
            # Any arguments to your chosen function go here

         QueryNorm:
            _target_:
            anemoi.models.layers.normalization.AutocastLayerNorm
            _partial_: True 
            bias: False

         KeyNorm:
            _target_:
            anemoi.models.layers.normalization.AutocastLayerNorm
            _partial_: True 
            bias: False

      encoder:
         LayerNorm:
            _target_: torch.nn.LayerNorm 
            _partial_: True

         Linear:
            _target_: torch.nn.Linear 
            _partial_: True

      decoder:
         LayerNorm:
            _target_: torch.nn.LayerNorm 
            _partial_: True

         Linear:
            _target_: torch.nn.Linear _partial_: True
```

File: models/docs/modules/models.rst [hunk 16]
```diff
@@ -2,40 +2,82 @@
...
+.. autoclass:: anemoi.models.models.hierarchical.AnemoiModelEncProcDecHierarchical
+   :members:
+   :no-undoc-members:
+   :show-inheritance:
+
+*************************
+ Time Interpolator Model
+*************************
+
+A specialized architecture for time interpolation tasks.
```
Comment (HCookie) [inline models/docs/modules/models.rst:77]:
```suggestion
A specialized architecture for time interpolation tasks.

Key features:

 #. Ability to select time indices for forcing and predictions
 
 #. Allows for provision of t0 and t6 and predictions of t1->5
```

File: training/docs/modules/strategy.rst [hunk 17]
```diff
@@ -12,5 +12,32 @@ level which can also run on multiple GPUs, and is a standard strategy
 within PyTorch `DDP Strategy
 <https://pytorch.org/tutorials/intermediate/ddp_tutorial.html>`__.
 
-Generally you should not need to change this module, as it is
-independent of the computer being used for training.
+.. note::
+
+   Generally you should not need to change this module, as it is
+   independent of the computer being used for training.
```
Comment (HCookie) [inline training/docs/modules/strategy.rst]:
```suggestion
   independent of the system being used for training.
```

**Review by HCookie — COMMENTED** [review]
Fantastic work, very much appreciate the continued support of the documentation.

File: models/docs/introduction/overview.rst (hunk 14 above)
Comment (ssmmnn11) [reply models/docs/introduction/overview.rst:69]: we could put e.g. AIFS-single and e.g. AIFS-CRPS instead. then it is clear where to look for more insight.

Comment (ssmmnn11) [reply models/docs/introduction/overview.rst]: > Might be a good idea to remove the reference to the AIFS here, and just leave it is a score optimised ensemble model?

we could put e.g. AIFS-single and e.g. AIFS-CRPS instead. then it is clear where to look for more insight.

diffusion models are also score based so we should maybe talk about crps directly.

File: training/docs/user-guide/models.rst (hunk 2 above)
Comment (jakob-schloer) [reply training/docs/user-guide/models.rst]: Changed to pre-processing.

File: training/docs/user-guide/training.rst (hunk 3 above)
Comment (jakob-schloer) [reply training/docs/user-guide/training.rst]: Changed to right formatting.

File: training/src/anemoi/training/losses/weightedloss.py (hunk 4 above)
Comment (jakob-schloer) [reply training/src/anemoi/training/losses/weightedloss.py]: I agree, I think I had to change the docstrings due to the pre-commit hooks for docs.

File: models/docs/introduction/overview.rst (hunk 6 above)
Comment (jakob-schloer) [reply models/docs/introduction/overview.rst]: Thanks for spotting this. I think the linter intervened here.

File: models/docs/introduction/overview.rst (hunk 9 above)
Comment (jakob-schloer) [reply models/docs/introduction/overview.rst]: I've changed the nomenslecture for the mapper and processor but didn't go through the whole docs.

File: training/docs/user-guide/training.rst (hunk 12 above)
Comment (jakob-schloer) [reply training/docs/user-guide/training.rst]: True, would be nice though :)

File: models/docs/introduction/overview.rst (hunk 14 above)
Comment (jakob-schloer) [reply models/docs/introduction/overview.rst]: I've changed it to: 
"**AnemoiEnsModelEncProcDec**: The CRPS-optimized ensemble version that injects noise in the processor, e.g. AIFS-CRPS"

File: models/docs/introduction/overview.rst (hunk 15 above)
Comment (jakob-schloer) [reply models/docs/introduction/overview.rst]: I've already changed that but thanks for fixing.

**Comment by jakob-schloer** [comment]
> Hi Jakob, thanks for updating the doc! I think from a user's point of view, it would be really useful to have a tutorial/user guide named "From deterministic to probabilistic" explaining what needs to be changed when moving from one scenario to the other.

I've added a user guide called "Ensemble CRPS-based training".

**Comment by mchantry** [comment]
This is an amazing contribution, the guide is a particular highlight

**Review by mchantry — APPROVED** [review]
Great contribution and thanks for incorporating all the feedback, this can be merged.
